### DISCUSSION TITLE
Sharing the test archive

### DISCUSSION THREAD

**Discussion by fcorbelli** [mhx/dwarfs#61] [General]
I noticed that, during the comparison with zpaq, the "placebo" compression mode (-m5) was used while, in reality, the default one (-m1) is almost always used.  
Could you please share the file you used for testing to make some analysis?  
Thanks

**Comment by mhx** [comment]
Hi! Thanks for the feedback! I've converted this into a discussion, as there's no real issue with the project itself.

I'll see if I can upload a tarball with the test data somewhere and will keep you posted.

**Comment by mhx** [comment]
That being said, I *think* I've used `-m5` simply because it was listed in the `zpaq` docs as offering the best compression ratio. Upon looking closer at the benchmark diagram, I now realise that the time axis is logarithmic and `-m5` is an order of magnitude slower than `-m4` (and two orders compared to `-m1`). I'll see if I can easily repeat the tests and will update the documentation accordingly.

**Comment by mhx** [comment]
FYI, I just checked with the default compression:
```
$ time zpaq a perl-install.default.zpaq install
[...]
2258234 +added, 0 -removed.

0.000000 + (51161.953159 -> 8932.000297 -> 1142.876164) = 1142.876164 MB
559.804 seconds (all OK)

real    9m19.832s
user    7m27.430s
sys     0m52.514s
```
So that's indeed pretty much exactly two orders of magnitude faster (in terms of user time) than `-m5` at the cost of the resulting archive being slightly more than double the size.

`-m4` seems to be somewhere in between and gives pretty good compression:
```
$ time zpaq a perl-install.m4.zpaq install -m4
[...]
2258234 +added, 0 -removed.

0.000000 + (51161.953159 -> 8932.000297 -> 609.073623) = 609.073623 MB
1044.178 seconds (all OK)

real    17m24.218s
user    186m6.035s
sys     1m42.729s
```

**Reply by fcorbelli** [reply]
Please pay attention to versioning: as written above try to modify a file inside the "source" folder.
And relaunch zpaq with the exact same parameters

PS and no, in 99% of the case you will use -m1 (the default), nor -m2, m3, m4 and m5

**Comment by fcorbelli** [comment]
zpaq (and forks :) have the ability to make deduplicated backup (aka: crontabbed) of very big file (virtual machine's vmdk) or very big but typically static folders (aka: a fileserver), or huge e-mail store server, MySQL dump backup and so on.
It is just like a "time machine" in a single file, that is always "append only" (aka: rsync --append will run blazing fast on zpaq's archive).  
In every file you can have some thousand different version of the same folders.
Short version: a data manager tool to retain backups... forever

The use is not so much "launch once on a folder, and that's it", but "launch a thousand times, on the same folder, at different moments"

In other zpaq's forks there is the possibility of performing parallel-hashed (multithreaded) file checks with actual speeds, real world, ~1-2GB/s, on non-spinning drives (SSDs and much better on NVMe)
And most of all, it works the same way on Windows, Linux, BSD-Unix, Solaris, Mac, QNAP nas (!), ESXi host

If I understand right dwarf (very good software indeed!) cannot update an image, therefore it is (almost) useless for daily, or every 60 minutes, backup purpose.
But can be **very** interesting for long-term archiving (once for year?)

I would really appreciate do some analysis and comparisons on your test-set archive

**Reply by mhx** [reply]
> zpaq (and forks :) have the ability to make deduplicated backup (aka: crontabbed) of very big file (virtual machine's vmdk) or very big but typically static folders (aka: a fileserver), or huge e-mail store server, MySQL dump backup and so on. It is just like a "time machine" in a single file, that is always "append only" (aka: rsync --append will run blazing fast on zpaq's archive). In every file you can have some thousand different version of the same folders. Short version: a data manager tool to retain backups... forever

I'm aware of this feature and it's pretty cool indeed!

> If I understand right dwarf (very good software indeed!) cannot update an image, therefore it is (almost) useless for daily, or every 60 minutes, backup purpose. But can be **very** interesting for long-term archiving (once for year?)

In fact, I've actually got some ideas for adding incremental updates to DwarFS as well. Also, because DwarFS images cannot be modified, this would be done by layering incremental update images over an initial "full" image. Storing multiple images in a single file should be easily possible. If (when) this works out, I imagine you'd be able to easily mount hundreds of different snapshots of your data simultaneously. Alas, I'd need some time to actually implement this. :)

> I would really appreciate do some analysis and comparisons on your test-set archive

Here's the [test data tarball](https://www.dropbox.com/s/vmwxr5wdenkg0pe/perl-install.tar.xz?dl=0). Let me know if that helps!

**Reply by fcorbelli** [reply]
Thank you. I just prepare dwarf on a ubuntu virtual machine.
Which filesystem (to store the image) do you suggest?
Straight-plain ext4?

**Reply by mhx** [reply]
Not sure if it matters much, especially on a VM. I've run all benchmarks on bare metal with XFS on pretty fast SSDs.

**Comment by fcorbelli** [comment]
In fact the key is here
```
51161.953159 -> 8932.000297
```
With a rolling-SHA-1 (you can change the default "chunk" size: the smaller, the better deduplication, but more time is needed) in the first stage the 51GB is shrinked to 8.9GB.  
Then compressed.  
But, in real world, sometimes you will use even... no compression at all (-m0) to store, for example, encrypted virtual volume (that cannot be compressed at all).

Short version: zpaq is NOT something like 7z or RAR (for example), it's something... fundamentally different (of course you can use as 7z: it runs just about as fast, and compress just about as 7z).
But the multiversion-snapshot-like is just another world

**Reply by mhx** [reply]
> With a rolling-SHA-1 (you can change the default "chunk" size: the smaller, the better deduplication, but more time is needed) in the first stage the 51GB is shrinked to 8.9GB.

Yeah, that the whole point of the test data set I've uploaded: it's *extremely* redundant, with lots of actual dupes and lots of files that are almost identical.

In my benchmark, DwarFS shrinks the 47.65 GiB (~51GB) to 4.26 GiB before even starting to use a "real" compression algorithm. One benefit of exploiting redundancy early is that you need to compress fewer blocks, which saves a lot of time. Another benefit is that more data can be held in a cache of the same size when the filesystem is mounted.

**Comment by fcorbelli** [comment]
Cannot compare on bare metal (dwarf does not run on BSD or Windows, so a ubuntu virtual machine is used, 16 CPU, 16GB, ~2GB/s media).  
Therefore speed compare is not really sure, too many things can disturb.  

However ... the testbed: 3 small virtual machines
```
root@utente-virtual-machine:/mnt/disk2# zpaqfranz dir testvm /s
zpaqfranz v54.9a-experimental archiver,  compiled Oct 16 2021
==== Scanning dir <<testvm/>>  /s
2021-10-16 14:32:23 <DIR>               testvm/
2021-10-16 14:32:07 <DIR>               testvm/centos8/
2021-08-06 17:03:13               8.684 testvm/centos8/Centos.nvram
2021-08-06 17:03:13       5.546.901.504 testvm/centos8/Centos.vmdk
2021-06-17 18:00:44                   0 testvm/centos8/Centos.vmsd
2021-08-06 17:03:14               2.576 testvm/centos8/Centos.vmx
2021-06-17 18:00:44                 261 testvm/centos8/Centos.vmxf
2021-07-21 21:31:35             177.177 testvm/centos8/vmware-0.log
2021-06-17 18:22:12             209.177 testvm/centos8/vmware-1.log
2021-08-06 17:03:14             161.615 testvm/centos8/vmware.log
2021-10-16 14:32:23 <DIR>               testvm/debian/
2021-10-16 12:00:45               8.684 testvm/debian/deb.nvram
2021-10-16 12:00:45       7.113.867.264 testvm/debian/deb.vmdk
2021-07-21 13:52:43                   0 testvm/debian/deb.vmsd
2021-10-16 12:00:46               2.184 testvm/debian/deb.vmx
2021-07-21 13:52:43                 258 testvm/debian/deb.vmxf
2021-09-07 15:36:17             151.112 testvm/debian/vmware-0.log
2021-09-05 14:25:57             158.203 testvm/debian/vmware-1.log
2021-08-29 13:11:36             154.698 testvm/debian/vmware-2.log
2021-10-16 12:00:46             150.271 testvm/debian/vmware.log
2021-10-16 14:23:05 <DIR>               testvm/fedora34/
2021-10-16 12:05:38       1.420.820.480 testvm/fedora34/Fedora 64-bit-000001.vmdk
2021-10-16 12:01:21              28.866 testvm/fedora34/Fedora 64-bit-Snapshot3.vmsn
2021-10-16 12:05:38               8.684 testvm/fedora34/Fedora 64-bit.nvram
2021-10-16 12:01:20       5.735.055.360 testvm/fedora34/Fedora 64-bit.vmdk
2021-10-16 12:01:21                 418 testvm/fedora34/Fedora 64-bit.vmsd
2021-10-16 12:05:40               3.453 testvm/fedora34/Fedora 64-bit.vmx
2021-10-16 14:23:05 <DIR>               testvm/fedora34/Fedora 64-bit.vmx.lck/
2021-10-16 12:00:54                 512 testvm/fedora34/Fedora 64-bit.vmx.lck/M45628.lck
2021-08-07 09:49:16                 268 testvm/fedora34/Fedora 64-bit.vmxf
2021-09-18 11:25:00              56.718 testvm/fedora34/mksSandbox-0.log
2021-08-10 14:07:16              57.735 testvm/fedora34/mksSandbox-1.log
2021-08-07 10:02:38              57.361 testvm/fedora34/mksSandbox-2.log
2021-10-16 12:05:38              56.946 testvm/fedora34/mksSandbox.log
2021-09-18 11:25:02             235.992 testvm/fedora34/vmware-0.log
2021-08-10 14:07:17             246.107 testvm/fedora34/vmware-1.log
2021-08-07 10:02:40             302.633 testvm/fedora34/vmware-2.log
2021-10-16 12:05:40             243.625 testvm/fedora34/vmware.log
2021-10-16 14:32:37 <DIR>               testvm/solaris/
2021-09-18 11:21:21               8.684 testvm/solaris/sol11.nvram
2021-09-18 11:21:21       6.611.664.896 testvm/solaris/sol11.vmdk
2021-08-31 12:09:42                   0 testvm/solaris/sol11.vmsd
2021-09-18 11:21:24               2.244 testvm/solaris/sol11.vmx
2021-08-31 12:09:42                 260 testvm/solaris/sol11.vmxf
2021-09-07 15:33:45             133.979 testvm/solaris/vmware-0.log
2021-09-05 14:25:51             148.099 testvm/solaris/vmware-1.log
2021-08-31 17:17:50             328.645 testvm/solaris/vmware-2.log
2021-09-18 11:21:24             152.070 testvm/solaris/vmware.log
        42 File          26.431.567.703 byte
         6 directory 0 bytes (    0.00 B) free

0.003 seconds (00:00:00)  (all OK)
```

dwarf with default options
```
root@utente-virtual-machine:/mnt/disk2# mkdwarfs -i testvm -o testvm.dwarf
I 16:36:27.120811 scanning testvm
I 16:36:27.126890 assigning directory and link inodes...
I 16:36:27.126954 waiting for background scanners...
I 16:38:30.301553 scanning CPU time: 393.9s
I 16:38:30.301649 finalizing file inodes...
I 16:38:30.302173 saved 0 B / 24.62 GiB in 2/42 duplicate files
I 16:38:30.302572 assigning device inodes...
I 16:38:30.302603 assigning pipe/socket inodes...
I 16:38:30.302637 building metadata...
I 16:38:30.302690 building blocks...
I 16:38:30.302719 saving names and symlinks...
I 16:38:30.303359 updating name and link indices...
I 16:38:30.304483 using a 4 KiB window at 512 B steps for segment analysis
I 16:38:30.304536 bloom filter size: 64 KiB
I 16:38:30.304781 ordering 40 inodes using nilsimsa similarity...
I 16:38:30.305337 nilsimsa: depth=20000 (1000), limit=255
I 16:38:30.305394 pre-sorted index (7 name, 0 path lookups) [4.26us]
I 16:38:30.305449 40 inodes ordered [623.3us, 408.5us CPU]
I 16:38:30.305524 waiting for segmenting/blockifying to finish...
I 16:49:37.796775 segmenting/blockifying CPU time: 179.3s
I 16:49:37.796861 bloom filter reject rate: 94.057% (TPR=0.046%, lookups=23077734151)
I 16:49:37.796904 segmentation matches: good=371166, bad=353967, total=2401014
I 16:49:37.797455 segmentation collisions: L1=0.081%, L2=0.031% [44995071 hashes]
I 16:49:37.797760 saving chunks...
I 16:49:37.805788 saving directories...
I 16:49:37.805954 saving shared files table...
I 16:49:37.813628 saving names table... [7.6ms]
I 16:49:37.813783 saving symlinks table... [2.58us]
I 16:49:37.936969 waiting for compression to finish...
I 16:49:42.555771 compressed 24.62 GiB to 10.11 GiB (ratio=0.410634)
I 16:49:42.557164 compression CPU time: 8090s
I 16:49:42.557271 filesystem created without errors [795.4s]
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
waiting for block compression to finish
6 dirs, 0/0 soft/hard links, 42/42 files, 0 other
original size: 24.62 GiB, dedupe: 0 B (2 files), segment: 3.156 GiB
filesystem: 21.46 GiB in 1374 blocks (524667 chunks, 40/40 inodes)
compressed filesystem: 1374 blocks/10.11 GiB written [depth: 20000]
```

795 s; now zpaqfranz with default options

```
root@utente-virtual-machine:/mnt/disk2# zpaqfranz a testvm.zpaq testvm
zpaqfranz v54.9a-experimental archiver,  compiled Oct 16 2021
Integrity check type: XXHASH64+CRC-32 + CRC-32 by fragments
Creating testvm.zpaq at offset 0 + 0
Adding 26.431.567.703 (24.62 GB) in 42 files  at 2021-10-16 14:51:09
 99.74% 00:00:00 (  24.55 GB) -> (  10.15 GB) of (  24.62 GB)   78.57 MB/sec
48 +added, 0 -removed.

0 + (26.431.567.703 -> 22.517.262.543 -> 10.972.475.386) = 10.972.475.386

325.773 seconds (000:05:25)  (all OK)
```


The archived files are just as big, but zpaqfranz is (in this case) more than 200% faster
```
root@utente-virtual-machine:/mnt/disk2# ls -l testvm.*
-rw-r--r-- 1 root root 10853694299 ott 16 16:49 testvm.dwarf
-rw-r--r-- 1 root root 10972475386 ott 16 16:56 testvm.zpaq
```


Mount and get xxh3 for every file (impressive speed!)

```
root@utente-virtual-machine:/mnt/disk2# dwarfs testvm.dwarf /mnt/caricato
I 17:02:19.515625 dwarfs (v0.5.6-9-g5112867, fuse version 35)
I 17:02:20.014080 file system initialized [498.4ms]
root@utente-virtual-machine:/mnt/disk2# cd /mnt
root@utente-virtual-machine:/mnt# cd caricato
root@utente-virtual-machine:/mnt/caricato# zpaqfranz sum * -xxh3 -all
zpaqfranz v54.9a-experimental archiver,  compiled Oct 16 2021
franz:-xxh3 
Getting XXH3 ignoring .zfs and :$DATA
                                                                           
Found (24.62 GB) => 26.431.567.703 bytes (24.62 GB) / 42 files in 0.009000

Creating 16 hashing thread(s) with XXH3
001% 00:00:21 ( 252.12 MB) of (  24.62 GB)          264.367.185 /sec
010% 00:00:16 (   2.46 GB) of (  24.62 GB)        2.643.192.913 /sec
020% 00:00:15 (   4.92 GB) of (  24.62 GB)        1.762.108.443 /sec
030% 00:00:14 (   7.38 GB) of (  24.62 GB)        1.321.587.213 /sec
040% 00:00:12 (   9.85 GB) of (  24.62 GB)        1.321.581.962 /sec
050% 00:00:10 (  12.31 GB) of (  24.62 GB)        1.321.578.811 /sec
060% 00:00:09 (  14.77 GB) of (  24.62 GB)        1.219.922.005 /sec
070% 00:00:07 (  17.23 GB) of (  24.62 GB)        1.156.382.405 /sec
080% 00:00:05 (  19.69 GB) of (  24.62 GB)        1.057.265.933 /sec
090% 00:00:02 (  22.15 GB) of (  24.62 GB)          991.185.462 /sec
XXH3: 099D7C10B1745DF71E14111FE315EE8D [            177.177]     centos8/vmware-0.log
XXH3: 20E6D6280C68B85B0B20486115E2CBB1 [            148.099]     solaris/vmware-1.log
XXH3: 38C2CE245133D2A0121FDEA58C0C3181 [                260]     solaris/sol11.vmxf
XXH3: 3D5F148B21CAD9718E5C6C302428BE3F [            243.625]     fedora34/vmware.log
XXH3: 43C43DB9AABBD1679FBCFB212C4F0C01 [            302.633]     fedora34/vmware-2.log
XXH3: 4982CDB8853CFE868022D2417B13B112 [              2.576]     centos8/Centos.vmx
XXH3: 4EDF03803878C98C0FCD1F31004212F3 [            161.615]     centos8/vmware.log
XXH3: 55A8EF4B92E9A360B19408255A92684D [              8.684]     debian/deb.nvram
XXH3: 5DEB921B33ADFBEDD8FDE9FAA1198451 [      7.113.867.264]     debian/deb.vmdk
XXH3: 67F3827AA285BDC37AF3383B1497DB92 [            235.992]     fedora34/vmware-0.log
XXH3: 6B24967AD9BE0B79C71FDEE4543E819A [      5.546.901.504]     centos8/Centos.vmdk
XXH3: 6E31DD76BEE96B9A3DBECD491BC41CFE [            209.177]     centos8/vmware-1.log
XXH3: 6F51087902860CECC817635DF0E2005E [             56.946]     fedora34/mksSandbox.log
XXH3: 70ACBEB98E354C45432923FB0C8ABE19 [            133.979]     solaris/vmware-0.log
XXH3: 7310F9C2B311C576AB6679E6614EBAFE [      5.735.055.360]     fedora34/Fedora 64-bit.vmdk
XXH3: 73A83C4FC6C302477291D2A598036CEE [              2.244]     solaris/sol11.vmx
XXH3: 7AEE35EE7DC40C45CC0F485AC1E6A4DA [             57.361]     fedora34/mksSandbox-2.log
XXH3: 87F47BB969B1448ACF6B9ED6F9BF5064 [              8.684]     centos8/Centos.nvram
XXH3: 900F882CEC5FBE520FB58B2DE26EACE7 [                268]     fedora34/Fedora 64-bit.vmxf
XXH3: 9765F4020A8A499BEAB39F37AE487568 [                512]     fedora34/Fedora 64-bit.vmx.lck/M45628.lck
XXH3: 99AA06D3014798D86001C324468D497F [                  0]     centos8/Centos.vmsd
XXH3: 99AA06D3014798D86001C324468D497F [                  0] === debian/deb.vmsd
XXH3: 99AA06D3014798D86001C324468D497F [                  0] === solaris/sol11.vmsd
XXH3: A22A8A22FA78A103B3DF5DE8EDD32120 [      6.611.664.896]     solaris/sol11.vmdk
XXH3: ADA22FF536E5A7AA837C2AC7E7234543 [              8.684]     fedora34/Fedora 64-bit.nvram
XXH3: AE9B6AC6D341192FA8AD8A29B2EBBB66 [              8.684]     solaris/sol11.nvram
XXH3: B06128B7CE5FEE0EFF936DB4F9B3491F [             57.735]     fedora34/mksSandbox-1.log
XXH3: B078CBE18FEAD10BD26E03A721B1A90D [            151.112]     debian/vmware-0.log
XXH3: B28F8F78F7B1A3730DBDC2690FEE8DFA [            246.107]     fedora34/vmware-1.log
XXH3: B37FABB14202E2507D76ACD577293399 [              2.184]     debian/deb.vmx
XXH3: B611FF7DA672A2907DE7B540C573AED0 [      1.420.820.480]     fedora34/Fedora 64-bit-000001.vmdk
XXH3: BA26F9FEC0FD44ED4815678FCB2F8D22 [            152.070]     solaris/vmware.log
XXH3: CBB9131DCFDED772AEC12394AF3C7193 [            328.645]     solaris/vmware-2.log
XXH3: CE4B2A73BFA35ACCB9CAC1650F9E70FC [             28.866]     fedora34/Fedora 64-bit-Snapshot3.vmsn
XXH3: CEBC204C79CD9F8E074469465D7AADC8 [                258]     debian/deb.vmxf
XXH3: DB8A1AED34B46BC0FDF4AA17D423688D [            154.698]     debian/vmware-2.log
XXH3: E1A0B67DECC5CC8455ED21AF020293E1 [            158.203]     debian/vmware-1.log
XXH3: E4FAD3C68359487DF0ECE92CA9207D6E [                261]     centos8/Centos.vmxf
XXH3: E8E1F719334F7404CE31A948A7177DE3 [            150.271]     debian/vmware.log
XXH3: EB67089D872590FE1B87173F8EBDCF94 [             56.718]     fedora34/mksSandbox-0.log
XXH3: F4204FF229F1CC0157B70BA0D9056600 [                418]     fedora34/Fedora 64-bit.vmsd
XXH3: FD30DFC5843CF227AC5CBDE16665F369 [              3.453]     fedora34/Fedora 64-bit.vmx

34.179 seconds (000:00:34)  (all OK)
```

Get xxh3 from "real" file (speed check against the "fake" dwarf)

```
root@utente-virtual-machine:/mnt/disk2# zpaqfranz sum testvm -xxh3 -all
zpaqfranz v54.9a-experimental archiver,  compiled Oct 16 2021
franz:-xxh3 
Getting XXH3 ignoring .zfs and :$DATA
                                                                           
Found (24.62 GB) => 26.431.567.703 bytes (24.62 GB) / 42 files in 0.002000

Creating 16 hashing thread(s) with XXH3
001% 00:00:15 ( 252.08 MB) of (  24.62 GB)          264.330.515 /sec
010% 00:00:14 (   2.46 GB) of (  24.62 GB)        2.643.221.779 /sec
020% 00:00:12 (   4.92 GB) of (  24.62 GB)        1.762.118.065 /sec
030% 00:00:11 (   7.38 GB) of (  24.62 GB)        1.982.371.652 /sec
040% 00:00:09 (   9.85 GB) of (  24.62 GB)        1.762.114.093 /sec
050% 00:00:08 (  12.31 GB) of (  24.62 GB)        1.651.977.122 /sec
060% 00:00:06 (  14.77 GB) of (  24.62 GB)        1.762.105.737 /sec
070% 00:00:05 (  17.23 GB) of (  24.62 GB)        1.682.013.599 /sec
080% 00:00:08 (  19.69 GB) of (  24.62 GB)          640.766.490 /sec
090% 00:00:03 (  22.15 GB) of (  24.62 GB)          679.669.331 /sec
XXH3: 099D7C10B1745DF71E14111FE315EE8D [            177.177]     testvm/centos8/vmware-0.log
XXH3: 20E6D6280C68B85B0B20486115E2CBB1 [            148.099]     testvm/solaris/vmware-1.log
XXH3: 38C2CE245133D2A0121FDEA58C0C3181 [                260]     testvm/solaris/sol11.vmxf
XXH3: 3D5F148B21CAD9718E5C6C302428BE3F [            243.625]     testvm/fedora34/vmware.log
XXH3: 43C43DB9AABBD1679FBCFB212C4F0C01 [            302.633]     testvm/fedora34/vmware-2.log
XXH3: 4982CDB8853CFE868022D2417B13B112 [              2.576]     testvm/centos8/Centos.vmx
XXH3: 4EDF03803878C98C0FCD1F31004212F3 [            161.615]     testvm/centos8/vmware.log
XXH3: 55A8EF4B92E9A360B19408255A92684D [              8.684]     testvm/debian/deb.nvram
XXH3: 5DEB921B33ADFBEDD8FDE9FAA1198451 [      7.113.867.264]     testvm/debian/deb.vmdk
XXH3: 67F3827AA285BDC37AF3383B1497DB92 [            235.992]     testvm/fedora34/vmware-0.log
XXH3: 6B24967AD9BE0B79C71FDEE4543E819A [      5.546.901.504]     testvm/centos8/Centos.vmdk
XXH3: 6E31DD76BEE96B9A3DBECD491BC41CFE [            209.177]     testvm/centos8/vmware-1.log
XXH3: 6F51087902860CECC817635DF0E2005E [             56.946]     testvm/fedora34/mksSandbox.log
XXH3: 70ACBEB98E354C45432923FB0C8ABE19 [            133.979]     testvm/solaris/vmware-0.log
XXH3: 7310F9C2B311C576AB6679E6614EBAFE [      5.735.055.360]     testvm/fedora34/Fedora 64-bit.vmdk
XXH3: 73A83C4FC6C302477291D2A598036CEE [              2.244]     testvm/solaris/sol11.vmx
XXH3: 7AEE35EE7DC40C45CC0F485AC1E6A4DA [             57.361]     testvm/fedora34/mksSandbox-2.log
XXH3: 87F47BB969B1448ACF6B9ED6F9BF5064 [              8.684]     testvm/centos8/Centos.nvram
XXH3: 900F882CEC5FBE520FB58B2DE26EACE7 [                268]     testvm/fedora34/Fedora 64-bit.vmxf
XXH3: 9765F4020A8A499BEAB39F37AE487568 [                512]     testvm/fedora34/Fedora 64-bit.vmx.lck/M45628.lck
XXH3: 99AA06D3014798D86001C324468D497F [                  0]     testvm/centos8/Centos.vmsd
XXH3: 99AA06D3014798D86001C324468D497F [                  0] === testvm/debian/deb.vmsd
XXH3: 99AA06D3014798D86001C324468D497F [                  0] === testvm/solaris/sol11.vmsd
XXH3: A22A8A22FA78A103B3DF5DE8EDD32120 [      6.611.664.896]     testvm/solaris/sol11.vmdk
XXH3: ADA22FF536E5A7AA837C2AC7E7234543 [              8.684]     testvm/fedora34/Fedora 64-bit.nvram
XXH3: AE9B6AC6D341192FA8AD8A29B2EBBB66 [              8.684]     testvm/solaris/sol11.nvram
XXH3: B06128B7CE5FEE0EFF936DB4F9B3491F [             57.735]     testvm/fedora34/mksSandbox-1.log
XXH3: B078CBE18FEAD10BD26E03A721B1A90D [            151.112]     testvm/debian/vmware-0.log
XXH3: B28F8F78F7B1A3730DBDC2690FEE8DFA [            246.107]     testvm/fedora34/vmware-1.log
XXH3: B37FABB14202E2507D76ACD577293399 [              2.184]     testvm/debian/deb.vmx
XXH3: B611FF7DA672A2907DE7B540C573AED0 [      1.420.820.480]     testvm/fedora34/Fedora 64-bit-000001.vmdk
XXH3: BA26F9FEC0FD44ED4815678FCB2F8D22 [            152.070]     testvm/solaris/vmware.log
XXH3: CBB9131DCFDED772AEC12394AF3C7193 [            328.645]     testvm/solaris/vmware-2.log
XXH3: CE4B2A73BFA35ACCB9CAC1650F9E70FC [             28.866]     testvm/fedora34/Fedora 64-bit-Snapshot3.vmsn
XXH3: CEBC204C79CD9F8E074469465D7AADC8 [                258]     testvm/debian/deb.vmxf
XXH3: DB8A1AED34B46BC0FDF4AA17D423688D [            154.698]     testvm/debian/vmware-2.log
XXH3: E1A0B67DECC5CC8455ED21AF020293E1 [            158.203]     testvm/debian/vmware-1.log
XXH3: E4FAD3C68359487DF0ECE92CA9207D6E [                261]     testvm/centos8/Centos.vmxf
XXH3: E8E1F719334F7404CE31A948A7177DE3 [            150.271]     testvm/debian/vmware.log
XXH3: EB67089D872590FE1B87173F8EBDCF94 [             56.718]     testvm/fedora34/mksSandbox-0.log
XXH3: F4204FF229F1CC0157B70BA0D9056600 [                418]     testvm/fedora34/Fedora 64-bit.vmsd
XXH3: FD30DFC5843CF227AC5CBDE16665F369 [              3.453]     testvm/fedora34/Fedora 64-bit.vmx

38.615 seconds (000:00:38)  (all OK)
```

OK now compare everything (check the global SHA256)
```
root@utente-virtual-machine:/mnt# zpaqfranz sum caricato -xxh3 -all -summary -noeta
zpaqfranz v54.9a-experimental archiver,  compiled Oct 16 2021
franz:-xxh3 -summary 
Getting XXH3 ignoring .zfs and :$DATA
                                                                           
Scanning filesystem time            0.010 s
Data transfer+CPU   time           33.244 s
Data output         time            0.000 s
Total size                     26.431.567.703 (  24.62 GB)
Tested size                    26.431.567.703 (  24.62 GB)
Duplicated size                             0 (    0.00 B)
Duplicated files                            2
Worked on 26.431.567.703 bytes avg speed (hashtime) 795.077.839 B/s
GLOBAL SHA256: CB01B10841F677B8080C821B609E5D5754F3916BADFF648F87D6015E57FC1C76

33.253 seconds (000:00:33)  (all OK)
```

... againts the "real" filesystem
```
root@utente-virtual-machine:/mnt/disk2# zpaqfranz sum testvm -xxh3 -all -summary -noeta
zpaqfranz v54.9a-experimental archiver,  compiled Oct 16 2021
franz:-xxh3 -summary 
Getting XXH3 ignoring .zfs and :$DATA
                                                                           
Scanning filesystem time            0.000 s
Data transfer+CPU   time           40.005 s
Data output         time            0.000 s
Total size                     26.431.567.703 (  24.62 GB)
Tested size                    26.431.567.703 (  24.62 GB)
Duplicated size                             0 (    0.00 B)
Duplicated files                            2
Worked on 26.431.567.703 bytes avg speed (hashtime) 660.706.604 B/s
GLOBAL SHA256: CB01B10841F677B8080C821B609E5D5754F3916BADFF648F87D6015E57FC1C76

40.004 seconds (000:00:40)  (all OK)
```

The "dwarf" image is equal to source




Short version: 
very, very good for fast extracting of "something" static
very good speed and size

**Comment by Phantop** [comment]
This is just me jumping in with unrelated stuff but I've got my own example of very redundant test data: every single version of Minecraft put through `precomp` (meaning that the jar files and the pngs within them are found as uncompressed data. Considering that I don't realistically intend on ever accessing this data, I've used a very large block size (128MiB) just to see how small I can get it. For reference, the source folder was about 20GiB before being put through `precomp`, 56.5GiB afterwards, and the compressed archive is a mere 900MiB at the end.

[Here it is.](https://drive.google.com/file/d/1VHafRa3qVkPPsaeEhDMuvjf9gHGz7Bf3/view?usp=sharing)

I've done some further compression of the "uncompressed" dwarfs file of that (the file I've shared is zstd compressed, that other version goes to about 11GiB) using `lrzip` and other algorithms and have gotten it down to almost half the size dwarfs with zstd alone gets it, of course without mounting or nearly as easy file access.
